server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -0,0 +1,424 @@
+                 fieldConstructor' . doNullability nullability <$> remoteSchemaInputObject schemaDoc defn
+               G.TypeDefinitionUnion _ -> throw400 RemoteSchemaError "expected input type, but got output type"
+               G.TypeDefinitionInterface _ -> throw400 RemoteSchemaError "expected input type, but got output type"
+       G.TypeList nullability subType -> buildField subType (fieldConstructor' . doNullability nullability . void . P.list)
+  in buildField fieldType fieldConstructor
+
+argumentsParser
+  :: forall n m
+  .  (MonadSchema n m, MonadError QErr m)
+  => G.ArgumentsDefinition
+  -> G.SchemaIntrospection
+  -> m (InputFieldsParser n ())
+argumentsParser args schemaDoc =
+  void . sequenceA <$> traverse (inputValueDefinitionParser schemaDoc) args
+
+-- | 'remoteField' accepts a 'G.TypeDefinition' and will returns a 'FieldParser' for it.
+--   Note that the 'G.TypeDefinition' should be of the GraphQL 'Output' kind, when an
+--   GraphQL 'Input' kind is provided, then error will be thrown.
+remoteField
+  :: forall n m
+   . (MonadSchema n m, MonadError QErr m)
+  => SchemaIntrospection
+  -> G.Name
+  -> Maybe G.Description
+  -> G.ArgumentsDefinition
+  -> G.TypeDefinition [G.Name]
+  -> m (FieldParser n ()) -- TODO return something useful, maybe?
+remoteField sdoc fieldName description argsDefn typeDefn = do
+  -- TODO add directives
+  argsParser <- argumentsParser argsDefn sdoc
+  case typeDefn of
+    G.TypeDefinitionObject objTypeDefn -> do
+      remoteSchemaObj <- remoteSchemaObject sdoc objTypeDefn
+      pure $ () <$ P.subselection fieldName description argsParser remoteSchemaObj
+    G.TypeDefinitionScalar (G.ScalarTypeDefinition desc name' _) ->
+      P.selection fieldName description argsParser <$> remoteFieldScalarParser name' desc
+    G.TypeDefinitionEnum enumTypeDefn ->
+      pure $ P.selection fieldName description argsParser $ remoteFieldEnumParser enumTypeDefn
+    G.TypeDefinitionInterface ifaceTypeDefn -> do
+      remoteSchemaObj <- remoteSchemaInterface sdoc ifaceTypeDefn
+      pure $ () <$ P.subselection fieldName description argsParser remoteSchemaObj
+    G.TypeDefinitionUnion unionTypeDefn -> do
+      remoteSchemaObj <- remoteSchemaUnion sdoc unionTypeDefn
+      pure $ () <$ P.subselection fieldName description argsParser remoteSchemaObj
+    _ -> throw400 RemoteSchemaError "expected output type, but got input type"
+
+remoteFieldScalarParser
+  :: forall m n
+   . (MonadParse n, MonadError QErr m)
+  => G.Name
+  -> Maybe G.Description
+  -> m (Parser 'Both n ())
+remoteFieldScalarParser name description =
+  case G.unName name of
+    "Boolean" -> pure $ P.boolean $> ()
+    "Int" -> pure $ P.int $> ()
+    "Float" -> pure $ P.float $> ()
+    "String" -> pure $ P.string $> ()
+    -- TODO IDs are allowed to be numbers, I think. But not floats. See
+    -- http://spec.graphql.org/draft/#sec-ID
+    "ID" -> pure $ P.string $> ()
+    _ -> pure $ P.unsafeRawScalar name description $> ()
```

Comment:
This function doesn't need to be in `m`, since all the branches use `pure`. Furthermore, use the name constants defined in `Hasura.RQL.Types.CustomTypes` rather than new string literals.

Reply:
Well we can't simply replace the constants here with their definition in `CustomTypes` because they haven't been defined as patterns.

Got rid of `m`.

Reply:
Ah yes, duh, sorry. Thanks!